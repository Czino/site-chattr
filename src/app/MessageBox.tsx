'use client'
import { useState } from 'react'
import { Textarea } from './Textarea'
import { currentDomain, currentURL } from './page'
import { usePostMessage } from './useConnectAndPost'

export const MessageBox = () => {
    const [content, setContent] = useState('')
    const postMessage = usePostMessage({ content, domain: currentDomain, url: currentURL })

    return (
        <>
            <Textarea id="textarea" value={content} onTextChange={setContent} placeholder="Your message..." />
            <button
                className="bg-gradient-purple text-white p-2 rounded hover:bg-gradient-purple-dark"
                onClick={postMessage}
            >
                Send
            </button>
        </>
    )
}
