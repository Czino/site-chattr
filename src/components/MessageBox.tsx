import { useState } from 'react'
import { IoMdSend } from 'react-icons/io'
import { usePostMessage } from '../hooks/usePostMessage'
import { Button } from './Button'
import { Textarea } from './Textarea'

type Props = { url: string }
export const MessageBox = ({ url }: Props) => {
    const [content, setContent] = useState('')
    const clearContent = () => setContent('')
    const postMessage = usePostMessage({
        content,
        url,
        onSuccess: clearContent,
    })

    return (
        <div className="flex flex-col gap-2">
            <div className="flex flex-row flex-nowrap gap-2 items-end p-2">
                <Textarea
                    className="flex-shrink pr-12 w-full"
                    id="textarea"
                    value={content}
                    onInput={(e) => {
                        const target = e.target as HTMLTextAreaElement
                        target.style.height = 'auto'
                        target.style.height = `${target.scrollHeight}px`
                    }}
                    onTextChange={setContent}
                    placeholder="Your message..."
                />
                <Button
                    onClick={postMessage}
                    aria-label="Send"
                    className="flex absolute right-3 bottom-3 justify-center items-center p-0 w-10 h-10 rounded-full"
                >
                    <IoMdSend className="w-7 h-7 relative left-[2px] flex-shrink-0" />
                </Button>
            </div>
        </div>
    )
}
