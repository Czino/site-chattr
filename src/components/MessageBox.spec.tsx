import NDK, { NDKEvent } from '@nostr-dev-kit/ndk'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { event1 } from '../../test/data/eventData'
import { MessageBox } from './MessageBox'

const ndk = new NDK()
const mockEvent = new NDKEvent(ndk, event1)
mockEvent.sign = jest.fn()
mockEvent.publish = jest.fn()
jest.mock('nostr-hooks', () => ({
    useNewEvent: () => ({
        createNewEvent: () => mockEvent,
    }),
}))
describe('MessageBox', () => {
    const url = 'https://localhost:3000/'
    const base = render(<MessageBox url={url} />).asFragment()
    it('should render correctly', () => {
        expect(base).toMatchSnapshot()
    })
    it('should publish event and clear text after', async () => {
        const message = 'my comment!'
        const { asFragment } = render(<MessageBox url={url} />)
        await userEvent.type(screen.getByPlaceholderText('Your message...'), message)
        expect(asFragment()).toMatchDiffSnapshot(base)
        await userEvent.click(screen.getByLabelText('Send'))
        expect(mockEvent.publish).toHaveBeenCalled()
        expect(mockEvent.content).toEqual(message)
        expect(asFragment()).toMatchDiffSnapshot(base)
    })
})
