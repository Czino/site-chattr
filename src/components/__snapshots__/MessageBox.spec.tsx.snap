// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageBox should publish event and clear text after 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,13 +1,11 @@
  <DocumentFragment>
    <textarea
      class=\\"resize-none border border-gray-300 rounded-md p-2 flex-shrink-0 focus:outline-none focus:ring-2 focus:ring-blue-500\\"
      id=\\"textarea\\"
      placeholder=\\"Your message...\\"
-   >
-     my comment!
-   </textarea>
+   />
    <button
      class=\\"bg-gradient-purple text-white p-2 rounded hover:bg-gradient-purple-dark \\"
    >
      Send
    </button>"
`;

exports[`MessageBox should publish event and clear text after 2`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`MessageBox should render correctly 1`] = `
<DocumentFragment>
  <textarea
    class="resize-none border border-gray-300 rounded-md p-2 flex-shrink-0 focus:outline-none focus:ring-2 focus:ring-blue-500"
    id="textarea"
    placeholder="Your message..."
  />
  <button
    class="bg-gradient-purple text-white p-2 rounded hover:bg-gradient-purple-dark "
  >
    Send
  </button>
</DocumentFragment>
`;
